File: week02-day3/class2-class4/ rqt/源码/src/sine_talker.cpp
#include "ros/ros.h"
#include "std_msgs/Float64.h"
#include <iostream>
#include <sys/time.h>
#include <math.h>

using namespace std;

unsigned long GetTime()
{
	struct timespec time1 = {0, 0};   
    clock_gettime(CLOCK_REALTIME, &time1); 
    long nows = time1.tv_nsec+ time1.tv_sec *1000000000;
    return nows;
}

int main(int argc, char **argv)
{
	ros::init(argc,argv,"sine_talker");//初始化ros,传递参数，定义节点名称
	ros::NodeHandle n;
	ros::Publisher chatter_pub = n.advertise<std_msgs::Float64> ("sine_signal",10);
	ros::Rate loop_rate(20.0);
	long time_init = 0;
	long time_now = 0;
	time_init = GetTime();

	while(ros::ok()){
		time_now = GetTime();
		std_msgs::Float64 msg;
		double sine = sin(  (double)( time_now - time_init )/1000000000 );
		if (0.8 > abs(sine))
			ROS_INFO("Normal output:%lf", abs(sine));
		else if (0.8 < abs(sine) && 0.9 > abs(sine))
			ROS_WARN("Danger output:%lf", abs(sine));
		else if (0.9 < abs(sine))
			ROS_ERROR("Wrong output:%lf", abs(sine));
		msg.data = sine;
		chatter_pub.publish(msg);
		ros::spinOnce();
		loop_rate.sleep();
	}
	return 0;
}

File: week02-day3/class2-class4/ rqt/源码/scripts/sine_talker.py
#!/usr/bin/env python
import rospy
import math
from std_msgs.msg import Float64

def sine_talker():
    pub = rospy.Publisher('sine_signal',Float64,queue_size=5)
    rospy.init_node('sine_talker', anonymous=True)
    rate = rospy.Rate(20)

    while not rospy.is_shutdown():
        time_now = rospy.get_time()
        sine_signal = math.sin(time_now)
        if 0.8 > abs(sine_signal):
            out_put = "Normal output: %s" % sine_signal
            rospy.loginfo(out_put)  
        elif 0.8 < abs(sine_signal) and 0.9 > abs(sine_signal):
            out_put = "Dangerous output: %s" %sine_signal
            rospy.logwarn(out_put)
        elif 0.9 < abs(sine_signal):
            out_put = "Wrong output: %s" %sine_signal
            rospy.logerr(out_put)
        pub.publish(sine_signal)
        rate.sleep()

if __name__ == '__main__':
    try:
        sine_talker()
    except rospy.ROSInterruptException:
        pass